### ISSUE TITLE
Stateful LSTM in pure Tensorflow after training with Keras

### ISSUE THREAD

**Issue by hicham-eyeem** [fchollet/keras#6321]
Hi there, 

I am having an issue with the LSTM layer with the stateful mode on. The problem is that sess.run() doesn't seem to work properly with tensors from this layer, probably related to the state of the cell that is not updated. Also, it is not clear how to reset the state when switching to pure Tensorflow. I tried to use Tensorboard to find the corresponding state tensor names but without a success. 

Here's an example : 
Model definition :
```
batch_size = 1
rnn_input = Input(batch_shape=(batch_size,1,embedding_dimension), dtype='float32', name='rnn_input')
lstm_layer = LSTM(lstm_layer_size,  batch_input_shape=(batch_size, 1, embedding_dimension), stateful=True, name='rnn')
final_dense_layer = Dense(vocabulary_len, activation = 'softmax', bias=True, name='dense')
rnn_model = final_dense_layer(lstm_layer(rnn_input))
model = Model(rnn_input, rnn_model)
```

Keras prediction : 
```
model.reset_states()
sequence_length = 24
rnn_input = ...  #first word of the sequence
for _ in range(sequence_length):
    word_prob = model.predict(rnn_input)
   do stuff with word_prob and set rnn_input to current word
```
Outputs something like this which is correct 

> water flowing from faucet in garden

Same session but now with Tensorflow (even when using model.reset_states() before) :
```
sequence_length = 24
rnn_input = ...  #first word of the sequence
rnn_input_name = .... #name of the input tensor
output_tensor = mode.output
for _ in range(sequence_length):
    word_prob = sess.run(output_tensor, feed_dict={rnn_input_name: rnn_input})
   do stuff with word_prob and set rnn_input to current word
```

Outputs the same thing regardless the input which is wrong 

> a bouldering # a bouldering # a bouldering # a bouldering etc.

Thank you very much in advance.

**Comment by Denisevi4** [comment]
I'm also having similar problems. I *think you also need to feed the state for LSTM each time you run a batch and also get the state back.

Something like that:
state = ...initialize to zero...
for _ in range(sequence_length):
    word_prob, state = sess.run([output_tensor, out_state], feed_dict={rnn_input_name: rnn_input, input_state:state})

Let me know if you figure it out. I'm also interested in the solution

**Comment by hicham-eyeem** [comment]
@Denisevi4 That is true, we need to keep track of both states (h and c) but for that we need the name of the corresponding tensors. I found a way to get it like this : 

```
layer = <your rnn layer>
update_ops = []
for old_value, new_value in layer.updates:
    update_ops.append(tf.assign(old_value, new_value))

```
So now `update_ops ` contains the two state tensors (they actually contain the Assign op). So to reset the state, they should be set to zeros, and every time prediction is made for one word, they should be updated.
So you have something like this : 

```
# get names of assign ops corresponding to the states via update_ops above
lstm_state1, lstm_state2 = 'Assign_249:0', 'Assign_250:0'
lstm_state1_op = sess.graph.get_tensor_by_name(lstm_state1)
lstm_state2_op = sess.graph.get_tensor_by_name(lstm_state2)

# Reset state
zero_state = np.zeros((1,<LSTM dimension>),dtype='float32')
sess.run(lstm_state1_op, feed_dict={lstm_state1: zero_state})
sess.run(lstm_state2_op, feed_dict={lstm_state2: zero_state})

# Update the states with something like this after every prediction
state1 = sess.run(lstm_state1_op, feed_dict={'name of lstm input' : rnn_input})
```

It doesn't output exactly the same output as Keras, but very close to it. 
Hope it helps!

**Comment by dsp-ml** [comment]
Hi

I am facing a similar issue with using Stateful and Steteless CuDNN GRU

I trained my model in Keras with CuDNN GRU's using let's say 1000 time steps and stateful=False.

During inference, to keep the model real-time, I changed the time_steps to 1, and stateful=true to load the model and it's weights in Keras. I converted this model to tensorflow protobuf using convert_variables_to_constants() function to freeze the graph and write_graph() to write it to a protobuf file.

The problem I am facing currently is that I am getting same results whether or not stateful = True or False during inference. which means I am missing something while converting keras model to tensorflow model. Is there any special consideration needed with the conversion process involving stateful CuDNN GRU's?

While running tensorflow session, along with the output, do I need to get states of all the CuDNN GRU layers and pass in these states with the next input? If yes, how can I get these states from the graph. It's not clear from the nodes of cu_dnngru layer, which of it might contain state information.

Or is there something very fundamental I am missing here? Any help is very much appreciated.

Tensorflow version : 1.14.0
Keras version : 2.2.4

Thanks
